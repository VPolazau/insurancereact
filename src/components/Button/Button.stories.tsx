import { Button } from './Button';

import type { Meta, StoryObj } from '@storybook/react';

const meta: Meta<typeof Button> = {
    title: 'UI/Button',
    tags: ['autodocs'],
    component: Button,
    argTypes: {
        children: {
            type: 'string',
            description: 'Название',
        },
        className: {
            type: 'string',
            defaultValue: 'btn',
        },
        variant: {
            type: 'string',
            description: 'Вариант внешнего вида',
            defaultValue: 'primary',
            options: ['primary', 'secondary', 'link'],
            control: {
                type: 'radio',
            },
        },
    },
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Primary: Story = {
    args: {
        children: 'Primary',
        variant: 'primary',
        disabled: false,
    },
};
export const Secondary: Story = {
    args: {
        children: 'Secondary',
        variant: 'secondary',
        disabled: false,
    },
};
export const Link: Story = {
    args: {
        children: 'Link',
        variant: 'link',
        disabled: undefined,
    },
};
