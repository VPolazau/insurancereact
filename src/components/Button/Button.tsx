import React, { ReactNode, MouseEventHandler, FormEventHandler } from 'react';
import cn from 'classnames';

import styles from './styles.module.css';

interface IButton extends React.ComponentProps<'button'> {
    variant?: 'primary' | 'secondary' | 'secondaryM' | 'link';
    children?: ReactNode | string;
    disabled?: boolean;
    className?: string;
    onClick?: MouseEventHandler<HTMLButtonElement> | undefined;
    onSubmit?: FormEventHandler<HTMLButtonElement> | undefined;
}

export const Button = ({
    variant = 'primary',
    children = '',
    disabled = false,
    onClick,
    onSubmit,
    className,
}: IButton) => {
    const btnClass = cn(className, styles.btn, styles[variant]);
    return (
        <button className={btnClass} disabled={disabled} onClick={onClick} onSubmit={onSubmit} data-testid="button">
            {children}
        </button>
    );
};
