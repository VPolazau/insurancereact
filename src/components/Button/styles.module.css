.btn {
  box-sizing: border-box;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  border: none;
  border-radius: 4px;
  font-family: 'Nunito Sans', sans-serif;
  font-style: normal;
  line-height: 130%;
  cursor: pointer;
  gap: 10px;
  user-select: none;
}

.primary {
  padding: 11px 30px;
  height: 44px;
  background-color: rgba(34, 192, 98, 1);
  font-weight: 400;
  font-size: 17px;
  text-align: center;
  font-feature-settings: 'liga' off;
  color: rgba(252, 255, 254, 1);
}

.primary:hover {
  background-color: rgba(30, 177, 89, 1);
}

.primary:active {
  box-shadow: inset 1px 1px 4px  rgb(21, 125, 63);
}

.primary:disabled {
  background-color: rgba(215, 221, 218, 1);
  color: rgba(139, 139, 139, 1);
  cursor: auto;
}

.primary:disabled:active {
  box-shadow: none;
}

.secondary {
  padding: 5px 12px;
  border: 1px solid #22C062;
  font-size: 14px;
  background-color: #fff;
  font-weight: 400;
  color: rgba(31, 29, 29, 1);
  font-feature-settings: 'pnum' on, 'lnum' on, 'liga' off;
}

.secondary:hover {
  background: #E6FEEF;
} 

.secondary:disabled {
  border: 1px solid #D7DDDA;
  color: rgba(139, 139, 139, 1);
  cursor: auto;
}

.secondary:disabled:hover {
  background: #fff;
}

.secondaryM {
  height: 44px;
  padding: 11px 30px;
  border: 1px solid #22C062;
  font-size: 17px;
  background-color: #fff;
  font-weight: 400;
  color: rgba(31, 29, 29, 1);
  font-feature-settings: 'pnum' on, 'lnum' on, 'liga' off;
}

.secondaryM:hover {
  background: #E6FEEF;
} 

.secondaryM:disabled {
  border: 1px solid #D7DDDA;
  color: rgba(139, 139, 139, 1);
  cursor: auto;
}

.secondaryM:disabled:hover {
  background: #fff;
}

.link {
  background-color: #fff;
  border: 1px solid transparent;
  padding: 5px 12px;
  color: #1F1D1D;
  font-size: 14px;
  font-feature-settings: 'liga' off;
}

.link:hover {
  background-color: #efefef;
}

.link:active {
  background-color: #dddddd;
}