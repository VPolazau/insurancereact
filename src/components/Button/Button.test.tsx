import { render, screen } from '@testing-library/react';

import { Button } from './Button';

describe('Button', () => {
    it('Button-primary', () => {
        render(<Button variant="primary" children="button" onClick={() => {}} />);
        expect(screen.getByTestId('button').textContent).toBe('button');
        expect(screen.getByTestId('button').classList.contains('primary')).toBe(true);
        expect(screen.getByTestId('button')).toBeInTheDocument();
        expect(screen.getByTestId('button')).toMatchSnapshot();
    });
    it('Button-secondary', () => {
        render(<Button variant="secondary" children="button" onClick={() => {}} />);
        expect(screen.getByTestId('button').textContent).toBe('button');
        expect(screen.getByTestId('button').classList.contains('secondary')).toBe(true);
        expect(screen.getByTestId('button')).toBeInTheDocument();
        expect(screen.getByTestId('button')).toMatchSnapshot();
    });
    it('Button-primary-disabled', () => {
        render(<Button variant="primary" children="button" disabled={true} onClick={() => {}} />);
        expect(screen.getByTestId('button').textContent).toBe('button');
        expect(screen.getByTestId('button').classList.contains('primary')).toBe(true);
        // @ts-ignore
        expect(screen.getByTestId('button').disabled).toBe(true);
        expect(screen.getByTestId('button')).toBeInTheDocument();
        expect(screen.getByTestId('button')).toMatchSnapshot();
    });
    it('Button-secondary-disabled', () => {
        render(<Button variant="secondary" children="button" disabled={true} onClick={() => {}} />);
        expect(screen.getByTestId('button').textContent).toBe('button');
        expect(screen.getByTestId('button').classList.contains('secondary')).toBe(true);
        // @ts-ignore
        expect(screen.getByTestId('button').disabled).toBe(true);
        expect(screen.getByTestId('button')).toBeInTheDocument();
        expect(screen.getByTestId('button')).toMatchSnapshot();
    });
    it('Button-link', () => {
        render(<Button variant="link" children="button" onClick={() => {}} />);
        expect(screen.getByTestId('button').textContent).toBe('button');
        expect(screen.getByTestId('button').classList.contains('link')).toBe(true);
        expect(screen.getByTestId('button')).toBeInTheDocument();
        expect(screen.getByTestId('button')).toMatchSnapshot();
    });
});
