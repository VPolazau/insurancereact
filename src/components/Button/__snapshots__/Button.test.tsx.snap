// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Button-link 1`] = `
<button
  class="btn link"
  data-testid="button"
>
  button
</button>
`;

exports[`Button Button-primary 1`] = `
<button
  class="btn primary"
  data-testid="button"
>
  button
</button>
`;

exports[`Button Button-primary-disabled 1`] = `
<button
  class="btn primary"
  data-testid="button"
  disabled=""
>
  button
</button>
`;

exports[`Button Button-secondary 1`] = `
<button
  class="btn secondary"
  data-testid="button"
>
  button
</button>
`;

exports[`Button Button-secondary-disabled 1`] = `
<button
  class="btn secondary"
  data-testid="button"
  disabled=""
>
  button
</button>
`;
